fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAutoCompleteResultCursor.java

```diff
@@ -169,12 +170,325 @@ private void performLookup() throws IOException {
+     *  non-matching chunk of text from the suggestion to the
+     *  provided fragments list.
+     *  @param sb The {@code StringBuilder} to append to
+     *  @param text The text chunk to add
+     */
+    protected void addNonMatch(StringBuilder sb, String text) {
+        sb.append(text);
+    }
+
+    /** Called while highlighting a single result, to append
+     *  the whole matched token to the provided fragments list.
+     *  @param sb The {@code StringBuilder} to append to
+     *  @param surface The surface form (original) text
+     *  @param analyzed The analyzed token corresponding to the surface form text
+     */
+    protected void addWholeMatch(StringBuilder sb, String surface, String analyzed) {
+        sb.append("<b>");
+        sb.append(surface);
+        sb.append("</b>");
+    }
+
+    /** Called while highlighting a single result, to append a
+     *  matched prefix token, to the provided fragments list.
+     *  @param sb The {@code StringBuilder} to append to
+     *  @param surface The fragment of the surface form
+     *        (indexed during build, corresponding to
+     *        this match
+     *  @param analyzed The analyzed token that matched
+     *  @param prefixToken The prefix of the token that matched
+     */
+    protected void addPrefixMatch(StringBuilder sb, String surface, String analyzed, String prefixToken) {
+        // TODO: apps can try to invert their analysis logic
+        // here, e.g. downcase the two before checking prefix:
+        if (prefixToken.length() >= surface.length()) {
+            addWholeMatch(sb, surface, analyzed);
+            return;
+        }
+        sb.append("<b>");
+        sb.append(surface.substring(0, prefixToken.length()));
+        sb.append("</b>");
+        sb.append(surface.substring(prefixToken.length()));
+    }
+
+    public RecordCursor<IndexEntry> lookup() throws IOException {
+        // Determine the tokens from the query key
+        final boolean phraseQueryNeeded = query.startsWith("\"") && query.endsWith("\"");
+        final String searchKey = phraseQueryNeeded ? query.substring(1, query.length() - 1) : query;
+        List<String> tokens = new ArrayList<>();
+        final String prefixToken = getQueryTokens(searchKey, tokens);
+
+        IndexReader indexReader = getIndexReader();
+        Set<String> fieldNames = new HashSet<>();
+        indexReader.leaves().forEach(leaf -> leaf.reader().getFieldInfos().forEach(fieldInfo -> fieldNames.add(fieldInfo.name)));
+        fieldNames.remove(LuceneIndexMaintainer.PRIMARY_KEY_FIELD_NAME);
+        fieldNames.remove(LuceneIndexMaintainer.PRIMARY_KEY_SEARCH_NAME);
+
+        final Set<String> tokenSet = new HashSet<>(tokens);
+        Query finalQuery = phraseQueryNeeded
+                             ? buildQueryForPhraseMatching(fieldNames, tokens, prefixToken)
+                             : buildQueryForTermsMatching(fieldNames, tokenSet, prefixToken);
```

Comment:
If I understand correctly, you refactored the original code to only get the ```tokens``` in ```getQueryTokens``` and then build the ```finalQuery``` in separate methods, to avoid duplicate code. And you have to pass in a list for phrase query building and set for terms query building. This looks fine to me, but maybe it is worthy to have a comment for '''buildQueryForPhraseMatching``` method to explain why it needs a List but the ```buildQueryForTermsMatching``` uses Set?